File: WeatherInfo/app/src/main/java/com/example/weatherinfo/MainActivity.java
package com.example.weatherinfo;
import com.google.zxing.*;

import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.media.Image;
import android.os.Bundle;
import android.util.Log;
import android.util.Size;
import android.widget.ImageView;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.annotation.OptIn;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.ExperimentalGetImage;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import android.Manifest;
import android.widget.TextView;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.zxing.common.HybridBinarizer;

import java.util.concurrent.ExecutionException;

public class MainActivity extends AppCompatActivity {
    private static final int PERMISSION_REQUEST_CAMERA = 83854;
    private ImageView preview;
    ListenableFuture<ProcessCameraProvider> cameraProviderFuture;
    TextView current_city;
    String current_qr = null;
    YUVtoRGB translator = new YUVtoRGB();
    public static String scanQRImage(Bitmap bMap) {
        String contents = null;

        int[] intArray = new int[bMap.getWidth()*bMap.getHeight()];
        bMap.getPixels(intArray, 0, bMap.getWidth(), 0, 0, bMap.getWidth(), bMap.getHeight());

        LuminanceSource source = new RGBLuminanceSource(bMap.getWidth(), bMap.getHeight(), intArray);
        BinaryBitmap bitmap = new BinaryBitmap(new HybridBinarizer(source));

        Reader reader = new MultiFormatReader();
        try {
            Result result = reader.decode(bitmap);
            contents = result.getText();
        }
        catch (Exception e) {
            Log.e("QrTest", "Error decoding barcode", e);
        }
        return contents;
    }
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });
        preview = findViewById(R.id.preview);
        current_city = findViewById(R.id.current_city);

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[] {Manifest.permission.CAMERA},
                    PERMISSION_REQUEST_CAMERA);
        } else {
            initializeCamera();
        }
    }
    private void initializeCamera() {
        cameraProviderFuture = ProcessCameraProvider.getInstance(this);
        cameraProviderFuture.addListener(new Runnable() {
            @Override
            public void run() {
                try {
                    ProcessCameraProvider cameraProvider = cameraProviderFuture.get();
                    ImageAnalysis imageAnalysis = new ImageAnalysis.Builder()
                            .setTargetResolution(new Size(1024, 768))
                            .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                            .build();

                    CameraSelector cameraSelector = new CameraSelector.Builder()
                            .requireLensFacing(CameraSelector.LENS_FACING_BACK)
                            .build();

                    imageAnalysis.setAnalyzer(ContextCompat.getMainExecutor(MainActivity.this), new ImageAnalysis.Analyzer() {
                        @OptIn(markerClass = ExperimentalGetImage.class)
                        @Override
                        public void analyze(@NonNull ImageProxy image) {
                            Image img = image.getImage();
                            Bitmap bitmap = translator.translateYUV(img, MainActivity.this);
                            preview.setRotation(image.getImageInfo().getRotationDegrees());
                            preview.setImageBitmap(bitmap);

                            current_qr = scanQRImage(bitmap);

                            if (current_qr != null)
                            {
                                current_city.setText(current_qr);
                            }
                            image.close();
                        }
                    });
                    cameraProvider.bindToLifecycle(MainActivity.this, cameraSelector, imageAnalysis);
                } catch (ExecutionException | InterruptedException e) {
                    throw new RuntimeException(e);
                }
            }
        }, ContextCompat.getMainExecutor(this));
    }
}

File: Beetles/app/src/main/java/com/example/beetles/MainActivity.kt
package com.example.beetles

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.media.MediaPlayer
import android.os.Bundle
import android.widget.Button
import android.widget.SeekBar
import android.widget.TextView
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

@Suppress("DEPRECATION")
class MainActivity : AppCompatActivity() {

    lateinit var sharedPreferences: SharedPreferences
    @SuppressLint("MissingInflatedId")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            return@setOnApplyWindowInsetsListener insets
        }
        sharedPreferences = getPreferences(Context.MODE_PRIVATE)


        var currentPlayerSettings: Settings


        if (intent.getSerializableExtra("settings") == null)
            if (sharedPreferences.contains("playerName") && sharedPreferences.contains("musicVolume") &&
                sharedPreferences.contains("soundVolume") && sharedPreferences.contains("beetlesMoveSpeed") &&
                sharedPreferences.contains("beetlesRespawnSpeed") && sharedPreferences.contains("beetlesMax")
            ) {
                currentPlayerSettings = Settings(
                    sharedPreferences.getString("playerName", "") ?: "",
                    sharedPreferences.getBoolean("musicVolume", true),
                    sharedPreferences.getBoolean("soundVolume", true),
                    sharedPreferences.getInt("beetlesMoveSpeed", 100),
                    sharedPreferences.getInt("beetlesRespawnSpeed", 1000),
                    sharedPreferences.getInt("beetlesMax", 20)
                )
            } else currentPlayerSettings = Settings()
        else currentPlayerSettings = (intent.getSerializableExtra("settings") as? Settings)!!

        val playerName = findViewById<TextView>(R.id.player_name)
        val start_button = findViewById<Button>(R.id.start_button)
        val settings_button = findViewById<Button>(R.id.settings_button)
        val stats_button = findViewById<Button>(R.id.stats_button)
        val exit_button = findViewById<Button>(R.id.exit_button)


        playerName.text = currentPlayerSettings.playerName

        start_button.setOnClickListener {
            val intent = Intent(this, GameActivity::class.java)
            intent.putExtra("settings", currentPlayerSettings)
            startActivity(intent)
        }

        settings_button.setOnClickListener {
           val intent = Intent(this, SettingsActivity::class.java)
            intent.putExtra("settings", currentPlayerSettings)
            startActivity(intent)
        }

        stats_button.setOnClickListener {
            val intent = Intent(this, StatsActivity::class.java)
            intent.putExtra("settings", currentPlayerSettings)
            startActivity(intent)
        }

        exit_button.setOnClickListener {
            finishAffinity()
            System.exit(0)
        }

        val editor = sharedPreferences.edit()
        editor.putString("playerName", currentPlayerSettings.playerName)
        editor.putBoolean("musicVolume", currentPlayerSettings.musicVolume)
        editor.putBoolean("soundVolume", currentPlayerSettings.soundVolume)
        editor.putInt("beetlesMoveSpeed", currentPlayerSettings.beetlesMoveSpeed)
        editor.putInt("beetlesRespawnSpeed", currentPlayerSettings.beetlesRespawnSpeed)
        editor.putInt("beetlesMax", currentPlayerSettings.beetlesMax)
        editor.apply()
    }
}